include/deal.II/matrix_free/matrix_free.h

```diff
@@ -255,22 +336,71 @@ class MatrixFree : public Subscriptor
+     * Option to control whether the indices stored in the DoFHandler
+     * should be read and the pattern for task parallelism should be
+     * set up in the initialize method of MatrixFree. The default
+     * value is true. Can be disabled in case the mapping should be
+     * recomputed (e.g. when using a deforming mesh described through
+     * MappingEulerian) but the topology of cells has remained the
+     * same.
      */
     bool                initialize_indices;
 
     /**
-     * Option to control whether the mapping information should be computed in
-     * the initialize method of MatrixFree. Defaults to true. Can be disabled
-     * when only some indices should be set up (e.g. when only a set of
-     * independent cells should be computed).
+     * Option to control whether the mapping information should be
+     * computed in the initialize method of MatrixFree. The default
+     * value is true. Can be disabled when only some indices should be
+     * set up (e.g. when only a set of independent cells should be
+     * computed).
      */
     bool                initialize_mapping;
+
+    /**
+     * Option to control whether the loops should overlap communications and
+     * computations as far as possible in case the vectors passed to the loops
+     * support non-blocking data exchange. In most situations, overlapping is
+     * faster in case the amount of data to be sent is more than a few
+     * kilobytes. If less data is sent, the communication is latency bound on
+     * most clusters (point-to-point latency is around 1 microsecond on good
+     * clusters by 2016 standards). Depending on the MPI implementation and
+     * the fabric, it may be faster to not overlap and wait for the data to
+     * arrive. The default is true, i.e., communication and computation are
+     * overlapped.
+     **/
+    bool                overlap_communication_computation;
+
+    /**
+     * By default, the face part will only hold those faces (and ghost
+     * elements behind faces) that are going to be processed locally. In case
+     * MatrixFree should have access to all neighbors on locally owned cells,
+     * this option enables adding the respective faces at the end of the face
+     * range.
+     **/
+    bool                hold_all_faces_to_owned_cells;
+
+    /**
+     * This data structure allows to assign a fraction of cells to different
+     * categories when building the information for vectorization. It is used
+     * implicitly when working with hp adaptivity but can also be useful in
+     * other contexts, such as in local time stepping where one would like to
+     * control which elements together form a batch of cells.
+     *
+     * This array is accessed by the number given by cell->active_cell_index()
+     * when working on the active cells with @p level_mg_handler set to -1 and
+     * by cell->index() for the level cells.
+     */
+    std::vector<unsigned int> cell_vectorization_category;
```

Comment:
is this in public of `AdditionalData`? So you just access it as `data.cell_vectorization_category[i] = 0;` etc? 
what happens if the vector is not initialized? I guess the size is `tria.n_locally_active_cells` (for fine level).

I guess this should indeed fix https://github.com/dealii/dealii/issues/5197 and allow grouping of cells.

Reply:
btw, do you have a test for this feature? Probably a two-cell tria should be enough to test that n_macro_cells ==2.

Reply:
The vector needs to be initialized by the user. I will add this to the documentation.

Regarding the test: Have a look at `tests/matrix_free/cell_categorization.cc`, https://github.com/dealii/dealii/pull/6300/files#diff-b27c3c05e4477363e999534f14beaee0 (not sure if the link survives rebasing I will do in a few minutes).